piccolo_api/mfa/authenticator/tables.py

```diff
@@ -0,0 +1,270 @@
+        :param db_encryption_key:
+            The secret is encrypted in the database - ``db_encryption_key``
+            can be any reasonably random string. It provides a little extra
+            protection vs storing the secret in plain text.
+        :param recovery_code_count:
+            How many recovery codes to generate for the user - this allows
+            them to still gain access if their phone is lost.
+
+        """
+        # Generate recovery codes
+
+        recovery_codes = [
+            generate_recovery_code() for _ in range(recovery_code_count)
+        ]
+
+        #######################################################################
+        # Hash the recovery codes
+
+        # Use the hashing logic from BaseUser.
+        # We want to use the same salt for all of the user's recovery codes,
+        # otherwise logging in using a recovery code will take a long time.
+        salt = BaseUser.get_salt()
+
+        hashed_recovery_codes = [
+            BaseUser.hash_password(password=recovery_code, salt=salt)
+            for recovery_code in recovery_codes
+        ]
+
+        #######################################################################
+        # Generate a shared secret
+
+        secret = cls.generate_secret()
+
+        # We'll encrypt the secret for storing in the database.
+        encrypted_secret = _encrypt(
+            value=secret, db_encryption_key=db_encryption_key
+        )
+
+        #######################################################################
+
+        instance = cls(
+            {
+                cls.user_id: user_id,
+                cls.secret: encrypted_secret,
+                cls.recovery_codes: hashed_recovery_codes,
+            }
+        )
+        await instance.save()
+
+        return (instance, recovery_codes)
+
+    @classmethod
+    async def revoke_all(cls, user_id: int):
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        await cls.update({cls.revoked_at: now}).where(
+            cls.user_id == user_id,
+            cls.revoked_at.is_null(),
+        )
+
+    @classmethod
+    async def authenticate(
+        cls, user_id: int, code: str, db_encryption_key: str
+    ) -> bool:
+        secret = (
+            await cls.objects()
+            .where(
+                cls.user_id == user_id,
+                cls.revoked_at.is_null(),
+            )
+            .order_by(cls.created_at, ascending=False)
+            .first()
+        )
+
+        if secret is None:
+            return False
+
+        pyotp = get_pyotp()
+
+        if secret.last_used_code == code:
+            logger.warning(
+                f"User {user_id} reused a token - potential replay attack."
+            )
+            return False
+
+        shared_secret = _decrypt(
+            encrypted_value=secret.secret, db_encryption_key=db_encryption_key
+        )
+        totp = pyotp.TOTP(shared_secret)  # type: ignore
+
+        if totp.verify(code):
```

Comment:
Consideration should be given to the `valid_window` argument in this method. This is fine as is, but it does slightly reduce usability if end users have a slight time delay between their clock and the server clock. Further, as this method will take a bit of time (miniscule but still) to do things such as decrypt the secret before verification it ever so slightly increases the chance that by the time MFA is verified it has expired. Bonus points if users submit the totp code within the last bit a code is valid on their phone or they are on the other side of the world to the application in question.

I will note that the current implementation is fine, and also means `last_used_code` is fine with a single `Text` entry. If Piccolo were to support a more conventional window (Current + one past + one future) then the `last_used_code` would need to essentially support three codes at once instead of one due to three codes being valid at any given time. Adding support for that does introduce a touch of complexity, but is something to consider from a usability side.

Here is an [RFC link](https://datatracker.ietf.org/doc/html/rfc6238#section-5.2) on validation and window sizes.

Again, feel free to resolve this point. Either of the two approaches is sufficient but I just wanted to raise it as a discussion point so you were aware of the implications.

Reply:
Yeah, it's tricky.

We could have a param to `MFAProvider` where the user can specify the `valid_window`.

With `last_used_code`, the logic does get quite tricky if we start storing multiple codes. It's to guard against replay attacks - presumably the vulnerability is if someone logs in twice in quick succession, and then if the first code is replayed it is still valid.

I'm not sure - what do you think?

Reply:
Yea, it is unfortunately one of those areas where I feel the cost to implement it is non linear haha.

Honestly for now it's probably just easier to leave as is, with a single valid token at any given time. If users do have issues they could also raise it and we could resolve it at a later date rather than adding that cost factor to this pr.

Both ways do work and are fairly equally valid choices I feel, in this case I'd probably err on the side of what's already implemented and just have it as a discussion point for possibly some time in the future

Reply:
OK, I'll add a `valid_window` argument for now, and will mention the pros and cons of increasing it. Thanks!